src/udp.rs

```diff
@@ -358,33 +362,29 @@ impl IsEmpty for (Vec<u8>, SocketAddr) {
     }
 }
 
+/// UDP socket receive result. It yields data only when socket returns EWOULDBLOCK.
+enum RecvResult<T> {
+    /// Stop reading socket and return some data, if any was read.
+    WouldBlock(Option<T>),
+    /// Keep reading UDP socket until it returns EWOULDBLOCK.
+    _ContinueRecv,
+}
+
+/// Buffers received UDP packets and checks when to stop calling `recv()`.
 fn handle_recv_res<T: IsEmpty>(
     recv_res: io::Result<T>,
     read_buffer: &mut VecDeque<T>,
-    would_block: &mut bool,
-) -> ::Res<Option<T>> {
+) -> ::Res<RecvResult<T>> {
     match recv_res {
         Ok(msg) => {
-            if msg.is_empty() {
-                let e = Err(SocketError::ZeroByteRead);
-                if !read_buffer.is_empty() {
-                    return match read_buffer.pop_front() {
-                        Some(msg) => Ok(Some(msg)),
-                        None => e,
-                    };
-                } else {
-                    return e;
-                }
+            if !msg.is_empty() {
+                read_buffer.push_back(msg);
```

Comment:
Hmm this is not what the reveiw comment was asking - will this not lead to unexpected errors (and maybe panics in examples or wrong decision being made in production) due to attempting to deserialise something that's empty ? See if read the [review comment](https://github.com/maidsafe/socket-collection/pull/20#discussion_r234262712) again and can spot what it was asking.

Reply:
Yep, this change does completely what you suggested: if message is empty, it does nothing, simply returns `Ok(RecvResult::ContinueRecv)`, if message has some bytes, it will enqueue those bytes for later reading, but it won't error if UDP payload was 0 bytes.

Reply:
Ah you are right !! My colour blindness made me miss the `!` sign in the `if` condition due to foreground-background colour combo, and i still missed it for like 5 secs while trying to make sense of your comment.